src/cephadm/cephadmlib/daemons/smb.py

```diff
@@ -0,0 +1,437 @@
+
+
+logger = logging.getLogger()
+
+
+class Features(enum.Enum):
+    DOMAIN = 'domain'
+    CLUSTERED = 'clustered'
+
+    @classmethod
+    def valid(cls, value: str) -> bool:
+        # workaround for older python versions
+        try:
+            cls(value)
+            return True
+        except ValueError:
+            return False
+
+
+class Config:
+    instance_id: str
+    source_config: str
+    samba_debug_level: int
+    debug_delay: int
+    domain_member: bool
+    clustered: bool
+    join_sources: List[str]
+    custom_dns: List[str]
+    smb_port: int
+    ceph_config_entity: str
+
+    def __init__(
+        self,
+        *,
+        instance_id: str,
+        source_config: str,
+        domain_member: bool,
+        clustered: bool,
+        samba_debug_level: int = 0,
+        debug_delay: int = 0,
+        join_sources: Optional[List[str]] = None,
+        custom_dns: Optional[List[str]] = None,
+        smb_port: int = 0,
+        ceph_config_entity: str = 'client.admin',
+    ) -> None:
+        self.instance_id = instance_id
+        self.source_config = source_config
+        self.domain_member = domain_member
+        self.clustered = clustered
+        self.samba_debug_level = samba_debug_level
+        self.debug_delay = debug_delay
+        self.join_sources = join_sources or []
+        self.custom_dns = custom_dns or []
+        self.smb_port = smb_port
+        self.ceph_config_entity = ceph_config_entity
+
+    def __str__(self) -> str:
+        return (
+            f'SMB Config[id={self.instance_id},'
+            f' source_config={self.source_config},'
+            f' domain_member={self.domain_member},'
+            f' clustered={self.clustered}]'
+        )
+
+
+class SambaContainerCommon:
+    def __init__(
+        self,
+        cfg: Config,
+    ) -> None:
+        self.cfg = cfg
+
+    def name(self) -> str:
+        raise NotImplementedError('samba container name')
+
+    def envs(self) -> Dict[str, str]:
+        cfg_uris = [self.cfg.source_config]
+        environ = {
+            'SAMBA_CONTAINER_ID': self.cfg.instance_id,
+            'SAMBACC_CONFIG': json.dumps(cfg_uris),
+        }
+        if self.cfg.ceph_config_entity:
+            environ['SAMBACC_CEPH_ID'] = f'name={self.cfg.ceph_config_entity}'
+        return environ
+
+    def envs_list(self) -> List[str]:
+        return [f'{k}={v}' for (k, v) in self.envs().items()]
+
+    def args(self) -> List[str]:
+        args = []
+        if self.cfg.samba_debug_level:
+            args.append(f'--samba-debug-level={self.cfg.samba_debug_level}')
+        if self.cfg.debug_delay:
+            args.append(f'--debug-delay={self.cfg.debug_delay}')
+        return args
+
+    def container_args(self) -> List[str]:
+        return []
+
+
+class SMBDContainer(SambaContainerCommon):
+    def name(self) -> str:
+        return 'smbd'
+
+    def args(self) -> List[str]:
+        return super().args() + ['run', 'smbd']
```

Comment:
for clarity, SMB is going to be all the same container image, just different ones (config, winbindd, smbd, mustjoin) get different args?

Reply:
Mostly true. There's a possibility we will want to bring in one other image for metrics.